### PR TITLE
Update `contains_table` to experimental row hasher and equality comparator

### PR DESCRIPTION
**Pull request by divyegala**
This is a part of #11844 

Benchmarks: https://github.com/rapidsai/cudf/pull/13119#issuecomment-1509335246

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by divyegala** [comment]
<detail>
<summary>benchmarks</summary>
<p>
Join<int32_t, int32_t>/left_anti_join_32bit/100000/100000/manual_time                                              +0.0261         +0.0266             0             0             0             0
Join<int32_t, int32_t>/left_anti_join_32bit/100000/400000/manual_time                                              +0.0575         +0.0519             0             0             0             0
Join<int32_t, int32_t>/left_anti_join_32bit/100000/1000000/manual_time                                             +0.0767         +0.0710             0             0             0             0
Join<int32_t, int32_t>/left_anti_join_32bit/10000000/10000000/manual_time                                          +0.0050         +0.0050            11            11            11            11
Join<int32_t, int32_t>/left_anti_join_32bit/10000000/40000000/manual_time                                          +0.0078         +0.0078            25            25            25            25
Join<int32_t, int32_t>/left_anti_join_32bit/10000000/100000000/manual_time                                         +0.0089         +0.0090            52            53            52            53
Join<int32_t, int32_t>/left_anti_join_32bit/100000000/100000000/manual_time                                        +0.0029         +0.0029           118           118           118           118
Join<int32_t, int32_t>/left_anti_join_32bit/80000000/240000000/manual_time                                         +0.0050         +0.0050           172           173           172           173
Join<int64_t, int64_t>/left_anti_join_64bit/50000000/50000000/manual_time                                          +0.0025         +0.0025            59            59            59            59
Join<int64_t, int64_t>/left_anti_join_64bit/40000000/120000000/manual_time                                         +0.0044         +0.0044            86            87            86            87
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/100000/100000/manual_time                                        +0.0405         +0.0360             0             0             0             0
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/100000/400000/manual_time                                        +0.0575         +0.0518             0             0             0             0
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/100000/1000000/manual_time                                       +0.0806         +0.0750             0             0             0             0
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/10000000/10000000/manual_time                                    +0.0412         +0.0412             5             5             5             5
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/10000000/40000000/manual_time                                    +0.0713         +0.0712            10            11            10            11
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/10000000/100000000/manual_time                                   +0.0862         +0.0861            21            23            21            23
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/100000000/100000000/manual_time                                  +0.0387         +0.0387            51            53            51            53
Join<int32_t, int32_t>/left_anti_join_32bit_nulls/80000000/240000000/manual_time                                   +0.0609         +0.0609            69            74            69            74
Join<int64_t, int64_t>/left_anti_join_64bit_nulls/50000000/50000000/manual_time                                    +0.0337         +0.0337            26            27            27            27
Join<int64_t, int64_t>/left_anti_join_64bit_nulls/40000000/120000000/manual_time                                   +0.0577         +0.0577            36            38            36            38
Join<int32_t, int32_t>/left_semi_join_32bit/100000/100000/manual_time                                              +0.0396         +0.0351             0             0             0             0
Join<int32_t, int32_t>/left_semi_join_32bit/100000/400000/manual_time                                              +0.0570         +0.0519             0             0             0             0
Join<int32_t, int32_t>/left_semi_join_32bit/100000/1000000/manual_time                                             +0.0695         +0.0641             0             0             0             0
Join<int32_t, int32_t>/left_semi_join_32bit/10000000/10000000/manual_time                                          +0.0049         +0.0048            11            11            11            11
Join<int32_t, int32_t>/left_semi_join_32bit/10000000/40000000/manual_time                                          +0.0077         +0.0077            25            25            25            25
Join<int32_t, int32_t>/left_semi_join_32bit/10000000/100000000/manual_time                                         +0.0090         +0.0090            52            53            52            53
Join<int32_t, int32_t>/left_semi_join_32bit/100000000/100000000/manual_time                                        +0.0029         +0.0029           118           118           118           118
Join<int32_t, int32_t>/left_semi_join_32bit/80000000/240000000/manual_time                                         +0.0048         +0.0048           171           172           171           172
Join<int64_t, int64_t>/left_semi_join_64bit/50000000/50000000/manual_time                                          +0.0025         +0.0024            59            59            59            59
Join<int64_t, int64_t>/left_semi_join_64bit/40000000/120000000/manual_time                                         +0.0042         +0.0042            86            86            86            86
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000/100000/manual_time                                        +0.0336         +0.0298             0             0             0             0
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000/400000/manual_time                                        +0.0575         +0.0522             0             0             0             0
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000/1000000/manual_time                                       +0.0808         +0.0755             0             0             0             0
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/10000000/10000000/manual_time                                    +0.0417         +0.0416             5             5             5             5
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/10000000/40000000/manual_time                                    +0.0697         +0.0695            10            11            10            11
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/10000000/100000000/manual_time                                   +0.0879         +0.0878            20            22            20            22
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/100000000/100000000/manual_time                                  +0.0392         +0.0392            51            53            51            53
Join<int32_t, int32_t>/left_semi_join_32bit_nulls/80000000/240000000/manual_time                                   +0.0636         +0.0636            69            73            69            73
Join<int64_t, int64_t>/left_semi_join_64bit_nulls/50000000/50000000/manual_time                                    +0.0344         +0.0343            26            27            26            27
Join<int64_t, int64_t>/left_semi_join_64bit_nulls/40000000/120000000/manual_time                                   +0.0586         +0.0586            35            37            35            37
</p>
</detail>

File: cpp/src/search/contains_table.cu [hunk 1]
```diff
@@ -408,31 +306,7 @@ rmm::device_uvector<bool> contains(table_view const& haystack,
...
-  // code path that flattens the input tables for row comparisons. This way is known to have
-  // better performance.
-  return contains_without_lists_or_nans(haystack, needles, compare_nulls, stream, mr);
-
-  // Note: We have to keep separate code paths because unifying them will cause performance
-  // regression for the input having no nested lists.
-  //
-  // TODO: We should unify these code paths in the future when performance regression is no longer
-  // happening.
+  return contains_impl(haystack, needles, compare_nulls, compare_nans, stream, mr);
```
Comment (ttnghia) [inline cpp/src/search/contains_table.cu:309]: If that function is called just once, please move its content here and remove it completely.

**Comment by GregoryKimball** [comment]
Thank you @divyegala for suggesting this change. We can just barely see a performance difference in the automated microbenchmarks - about 1.5% slower in the worst case. The difference is close to the noise level and acceptable in my opinion.

image

**Comment by abellina** [comment]
@divyegala can the 1.5% worst-case regression that @GregoryKimball found be attributed to any of the code changes?

**Comment by divyegala** [comment]
> @divyegala can the 1.5% worst-case regression that @GregoryKimball found be attributed to any of the code changes?

@abellina Yes, I would say that if it's not just noise and we consistently see a 1.5% worst-case regression then it's because of the code changes. That's where your help with the NDS query would be a good insight to have.

**Comment by abellina** [comment]
@divyegala my question was if you could explain why there would be a regression. e.g. would you expect a regression if there are nulls, certain data types, specific cuDF expressions.

**Comment by vyasr** [comment]
Based on what I see above, this PR should be ready for review then? No real performance concerns here IIUC.

**Comment by divyegala** [comment]
@vyasr yes, it's ready for review

**Comment by vyasr** [comment]
@abellina just in case you didn't see (I'm not sure if there was additional offline communication here) #11299 was the original performance regression that triggered the creation of the extra code paths that are removed in this PR (#11330 introduced the code).

File: cpp/src/search/contains_table.cu (hunk 1 above)
Comment (vyasr) [reply cpp/src/search/contains_table.cu:309]: Agreed, I don't see a need for a separate impl function here.

**Review by vyasr — APPROVED** [review]
Approving assuming Nghia's comment about combining functions is addressed.

**Review by ttnghia — CHANGES_REQUESTED** [review]
Since some code was removed, there are some headers such as struct utilities are no longer needed. Please clean them up too.

**Review by karthikeyann — COMMENTED** [review]
are there any need for additional unit tests for this change?

**Comment by divyegala** [comment]
> are there any need for additional unit tests for this change?

@karthikeyann this functionality already existed, this PR is just essentially a refactor

**Review by ttnghia — APPROVED** [review]

**Pull request merged**